### PR TITLE
feat: use private gitlab pypi registry together with gitlab host

### PR DESCRIPTION
**Pull request by deveaud-m**
## Changes:

Use Gitlab Pypi registries as private poetry sources to renovate projects hosted on the same Gitlab instance. 

## Context:

To configure poetry credentials for private pypi registries hosted on a Gitlab instance, I can create a `hostRules` entry like this:

```json
"hostRules": [
    {
      "matchHost": "https://gitlab.example.com/api/v4/projects/123/packages/pypi/simple/",
      "username": "gitlab-user",
      "password": "gitlab-token"
    }
]
```

However, if I want to renovate several projects using dependencies hosted on several registries, I would need a dedicated entry for each of those projects.

This PR add the possibility to use a dedicated hostRule only relevant for Gitlab hosted pypi registries. 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/poetry/artifacts.ts [hunk 1]
```diff
@@ -107,7 +107,8 @@ function getSourceCredentialVars(
   const envVars: Record<string, string> = {};
 
   for (const source of poetrySources) {
-    const matchingHostRule = find({ url: source.url });
+    const matchingHostRule =
+      find({ hostType: 'pypi', url: source.url }) || find({ url: source.url });
     const formattedSourceName = source.name.toUpperCase();
```
Comment (fgreinacher) [inline lib/manager/poetry/artifacts.ts:112]: While we're at it: This is also not completely right. According to [the Poetry documentaion](https://python-poetry.org/docs/master/configuration/#using-environment-variables):

> The environment variables must be prefixed by POETRY_ and are comprised of the uppercase name of the setting and **with dots and dashes replaced by underscore**

Should we also do that here or in a separate PR?

File: lib/manager/poetry/artifacts.ts [hunk 2]
```diff
@@ -107,7 +107,8 @@ function getSourceCredentialVars(
   const envVars: Record<string, string> = {};
 
   for (const source of poetrySources) {
-    const matchingHostRule = find({ url: source.url });
+    const matchingHostRule =
+      find({ hostType: 'pypi', url: source.url }) || find({ url: source.url });
```
Comment (viceice) [inline lib/manager/poetry/artifacts.ts:111]: What's the benefit of using a hostType here? The fallback should also match all possible hostType (so it ignore it).

**Review by viceice — CHANGES_REQUESTED** [review]
Please create an issue, so we can discuss the required changes. Also provide explanation what's not working today.

The current proposal don't look right to me.

**Comment by fgreinacher** [comment]
I'll jump in here because @deveaud-m is sick and we're a little bit blocked by this problem. Will create an issue that describes the situation.

**Comment by fgreinacher** [comment]
Issue is here: https://github.com/renovatebot/renovate/issues/14247

File: lib/manager/poetry/artifacts.ts (hunk 2 above)
Comment (fgreinacher) [reply lib/manager/poetry/artifacts.ts:111]: The problem with this lookup is that there might be other host rules that match the URL but are meant for something different. This is especially problematic when you want to/need to consume packages from the `platform` for which an host-rule is created automatically.

Comment (fgreinacher) [reply lib/manager/poetry/artifacts.ts:111]: We do a very similar thing for [NuGet](https://github.com/renovatebot/renovate/blob/b0ce30b59a2c31a58b2f71972b89ad13832cde9f/lib/manager/nuget/artifacts.ts#L42
), [Docker](https://github.com/renovatebot/renovate/blob/93d32806b4145041579494968f7f4966d0514461/lib/datasource/docker/index.ts#L273) or [Packagist](https://github.com/renovatebot/renovate/blob/93d32806b4145041579494968f7f4966d0514461/lib/datasource/packagist/index.ts#L32) datasources.

File: docs/usage/getting-started/private-packages.md [hunk 3]
```diff
@@ -327,6 +327,39 @@ For those found, a command similar to the following is run: `dotnet nuget add so
...
+`hostType`:
+
+```json
+{
+  "hostRules": [
+    {
+      "matchHost": "gitlab.example.com",
+      "hostType": "pypi",
+      "username": "gitlab-ci-token", // CI credentials will work on self-hosted renovate
+      "password": "$CI_JOB_TOKEN" // use any other username/password in other cases
```
Comment (viceice) [inline docs/usage/getting-started/private-packages.md:357]: This doesn't work. There is no environment interpolation in JSON.

File: lib/manager/poetry/artifacts.ts (hunk 2 above)
Comment (viceice) [reply lib/manager/poetry/artifacts.ts:111]: Ok so the PR description is misleading. You like to use a hostType to differentiate same hostname with different credentials in hostrules , so you don't need to have a lot more specific rules. 🤔

Can we update the PR description / issue requirements?
As in general we support passing credentials to poetry call.

Comment (viceice) [inline lib/manager/poetry/artifacts.ts:111]:
```suggestion
      find({ hostType: 'pypi', url: source.url }) ?? find({ url: source.url });
```

Will it make sense to use the datasource id for the hostType instead of a simple string here? So it will be future more safe.

File: lib/manager/poetry/artifacts.ts (hunk 1 above)
Comment (viceice) [reply lib/manager/poetry/artifacts.ts:112]: This should probably also fixed here, but can also be done in different PR.

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by fgreinacher** [comment]
@deveaud-m We can close this one in favor of #14255 😄

**Comment by deveaud-m** [comment]
Thanks @viceice for your comments and @fgreinacher for taking over!
